rust/operator-binary/src/spark_controller.rs

```diff
@@ -0,0 +1,968 @@
+                &rolegroup.role_group,
+            )
+            .build(),
+        spec: Some(ServiceSpec {
+            cluster_ip: Some("None".to_string()),
+            ports: Some(
+                build_ports(sc, rolegroup, rolegroup_config)
+                    .iter()
+                    .map(|(name, value)| ServicePort {
+                        name: Some(name.clone()),
+                        port: *value,
+                        protocol: Some("TCP".to_string()),
+                        ..ServicePort::default()
+                    })
+                    .collect(),
+            ),
+            selector: Some(role_group_selector_labels(
+                sc,
+                APP_NAME,
+                &rolegroup.role,
+                &rolegroup.role_group,
+            )),
+            publish_not_ready_addresses: Some(true),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+/// Build the [`StatefulSet`]s for the given rolegroup.
+///
+/// # Arguments
+/// * `_sc`                       - The cluster resource object.
+/// * `default_master_role_ports` - Master role service (and container ports). Used to build the master URLs needed by the worker pods.
+/// * `rolegroup`                 - The rolegroup.
+/// * `rolegroup_config`          - The validated configuration for the rolegroup.
+///
+fn build_rolegroup_statefulset(
+    sc: &SparkCluster,
+    default_master_role_ports: &[(String, i32)],
+    rolegroup_ref: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<StatefulSet, Error> {
+    match serde_yaml::from_str(&rolegroup_ref.role).unwrap() {
+        SparkRole::Master => build_master_stateful_set(sc, rolegroup_ref, rolegroup_config),
+        SparkRole::Worker => build_worker_stateful_set(
+            sc,
+            default_master_role_ports,
+            rolegroup_ref,
+            rolegroup_config,
+        ),
+        SparkRole::HistoryServer => build_history_stateful_set(sc, rolegroup_ref, rolegroup_config),
+    }
+}
+
+/// Build the [`StatefulSet`]s for the given [`SparkRole::Worker`] rolegroup.
+///
+/// # Arguments
+/// * `_sc`                       - The cluster resource object.
+/// * `default_master_role_ports` - Master role service (and container ports). Used to build the master URLs needed by the worker pods.
+/// * `rolegroup`                 - The rolegroup.
+/// * `rolegroup_config`          - The validated configuration for the rolegroup.
+///
+fn build_worker_stateful_set(
+    sc: &SparkCluster,
+    default_master_role_ports: &[(String, i32)],
+    rolegroup_ref: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<StatefulSet, Error> {
+    let rolegroup = sc
+        .spec
+        .workers
+        .as_ref()
+        .ok_or(Error::NoServerRole {
+            obj_ref: ObjectRef::from_obj(sc),
+        })?
```

Comment:
`ObjectRef` not required anymore and I would create a MissingRole error that takes a SparkRole?

Reply:
Changed to reference rolegroup. In the current implementation this cannot actually happen but this might change in the future.